Crashlytics/Crashlytics/Models/FIRCLSOnDemandModel.m

```diff
@@ -0,0 +1,209 @@
+                              stepDuration:(int)stepDuration {
+  self = [super init];
+  if (!self) {
+    return nil;
+  }
+
+  _uploadRate = uploadRate;
+  _base = base;
+  _stepDuration = stepDuration;
+
+  _dispatchQueue = dispatch_queue_create("com.google.firebase.crashlytics.on.demand", 0);
+  _queuedOperationsCount = 0;
+
+  _recordedOnDemandExceptionCount = 0;
+  _droppedOnDemandExceptionCount = 0;
+
+  _lastUpdated = [NSDate timeIntervalSinceReferenceDate];
+  _currentStep = -1;
+  _fileManager = [[NSFileManager alloc] init];
+
+  self.storedActiveReportPaths = [NSMutableArray array];
+
+  return self;
+}
+
+/*
+ * Called from FIRCrashlytics whenever the on-demand record exception method is called. Handles
+ * rate limiting and exponential backoff.
+ */
+- (BOOL)recordOnDemandExceptionIfQuota:(FIRExceptionModel *)exceptionModel
+             withDataCollectionEnabled:(BOOL)dataCollectionEnabled
+            usingExistingReportManager:(FIRCLSExistingReportManager *)existingReportManager {
+  // Record the exception model into a new report if there is unused on-demand quota. Otherwise,
+  // log the occurence but drop the event.
+  @synchronized(self) {
+    if ([self isQueueFull]) {
+      FIRCLSDebugLog(@"No available on-demand quota, dropping report");
+      [self getOrIncrementDroppedOnDemandEventCountForCurrentRun:YES];
+      return NO;  // Didn't record or submit the exception because no quota was available.
+    }
+
+    FIRCLSDataCollectionToken *dataCollectionToken = [FIRCLSDataCollectionToken validToken];
+    NSString *activeReportPath = FIRCLSExceptionRecordOnDemandModel(
+        exceptionModel, self.recordedOnDemandExceptionCount, self.droppedOnDemandExceptionCount);
+
+    if (!activeReportPath) {
+      FIRCLSErrorLog(@"Error recording on-demand exception");
+      return NO;  // Something went wrong when recording the exception, so we don't have a valid
+                  // path.
+    }
+
+    // Only submit an exception report if data collection is enabled. Otherwise, the report
+    // is stored until send or delete unsent reports is called.
+    [self incrementQueuedOperationCount:1];
+    [self getOrIncrementOnDemandEventCountForCurrentRun:YES];
+    dispatch_async(self.dispatchQueue, ^{
+      if (dataCollectionEnabled) {
+        [existingReportManager handleOnDemandReportUpload:activeReportPath
+                                      dataCollectionToken:dataCollectionToken
+                                                 asUrgent:YES];
+      } else {
+        [self.storedActiveReportPaths insertObject:activeReportPath atIndex:0];
+        if ([self.storedActiveReportPaths count] > FIRCLSMaxUnsentReports) {
+          [self.fileManager removeItemAtPath:[self.storedActiveReportPaths lastObject] error:nil];
+          [self.storedActiveReportPaths removeLastObject];
+          self->_recordedOnDemandExceptionCount -= 1;
+          self->_droppedOnDemandExceptionCount += 1;
```

Comment:
I think we can do `self.recordedOnDemandExceptionCount ...` for these

Reply:
they're read-only so we can't

Reply:
Or use the increment / decrement methods?